### PR TITLE
Add additional customization and configuration options for OpenTracing

### PR DESCRIPTION
**Pull request by hanwavefront**
Adding some nice-to-have features to make configuration and customization of tracing (mostly OpenTracing) easier:
1. ~~Add configuration property to set global tags for spans.~~ Will address this in separate PR.
2. Add configuration properties to configure sampling for OpenTracing, since that already exists for Sleuth.
3. Support overriding of the default WavefrontSpanReporter.
4. Add customizer for WavefrontTracer.Builder to allow for any further customizations.

### CODE REVIEW

**Comment by sushantdewan123** [comment]
lgtm

**Review by sushantdewan123 — APPROVED** [review]

**Comment by snicoll** [comment]
I'd like we move this to our next feature release. There are some changes in the current snapshot we need to push in a `2.0.2` first. I've flagged the one `on hold` accordingly.

File: wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java [hunk 1]
```diff
@@ -26,18 +35,50 @@
...
+    if (reporterArray.length > 1) {
+      compositeReporter = new CompositeReporter(reporterArray);
+    } else {
+      compositeReporter = reporterArray[0];
+    }
+    WavefrontTracer.Builder builder = new WavefrontTracer.Builder(compositeReporter, applicationTags)
         .excludeJvmMetrics();  // reported separately
     builder.redMetricsCustomTagKeys(
         new HashSet<>(wavefrontProperties.getTracing().getRedMetricsCustomTagKeys()));
+    builder.withGlobalTags(new HashMap<>(wavefrontProperties.getTracing().getTags()));
```
Comment (snicoll) [inline wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java:63]: I find this part confusing. `ApplicationTags` has `customTags` and we also have `globalTags`. What's the difference and why do we need two separate configuration options for this?

File: wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java [hunk 2]
```diff
@@ -26,18 +35,50 @@
...
+    List<Sampler> samplers =
+        createSampler(wavefrontProperties.getTracing().getOpenTracing().getSampler());
+    samplers.forEach(builder::withSampler);
+    customizers.orderedStream().forEach((customizer) -> customizer.customize(builder));
     return builder.build();
   }
 
+  private List<Sampler> createSampler(WavefrontProperties.Tracing.OpenTracing.Sampler samplerProperties) {
```
Comment (snicoll) [inline wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java:71]: could be named `createSamplers` considering the return type.

**Review by snicoll — COMMENTED** [review]
Thanks for the PR. I've made a few comments. 

The globalTags certainly seem it should have been a different commit and a different issue. Doing too many things in a single PR makes it hard to track individual changes and provide less visibility to what features were added in a release.

File: wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontProperties.java [hunk 3]
```diff
@@ -124,12 +127,23 @@ public void setExtractJvmMetrics(boolean extractJvmMetrics) {
...
     /**
      * Tags that should be associated with RED metrics. If the span has any of the
      * specified tags, then those get reported to generated RED metrics.
      */
     private Set<String> redMetricsCustomTagKeys = new HashSet<>();
 
+    /**
+     * Global tags included with every reported span.
+     */
+    private Map<String, String> tags = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (hanwavefront) [reply wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontProperties.java]: I think we can use the same tags for metrics and spans. Removed the property from this PR and will open a new PR to address.

File: wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontProperties.java [hunk 4]
```diff
@@ -124,12 +127,23 @@ public void setExtractJvmMetrics(boolean extractJvmMetrics) {
 
   public static class Tracing {
 
+    private final OpenTracing openTracing = new OpenTracing();
```
[the remark this answers is not in the document]
Comment (hanwavefront) [reply wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontProperties.java]: Fixed

File: wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java (hunk 2 above)
Comment (hanwavefront) [reply wavefront-spring-boot/src/main/java/com/wavefront/spring/autoconfigure/WavefrontTracingOpenTracingConfiguration.java]: Renamed

**Comment by snicoll** [comment]
@hanwavefront thanks for the update, I've polished things in 1047ab4, please take a look when you get a chance.

This PR has a number of changes which makes it hard to have a single description that the user can spot while browsing the changelog. Going forward, it would be better to have isolated PRs with more focused changes rather than a "theme" in a single PR.

The following improvements are available:

* A `WavefrontTracerBuilderCustomizer` callback interface to customize the auto-configured `WavefrontTracer`
* Configuration properties for OpenTracing's sampling (probability and duration)
* The default `Reporter` can be overridden and all such beans in the context are wrapped in a composite automatically

**Pull request closed without merging**